### PR TITLE
Refactor online detectors

### PR DESCRIPTION
**Pull request by ascillitoe**
This PR contains minor(ish?)  refactorings to the online drift detectors. A number of these are intended to be a stepping stone to implementing the non-parametric drift detectors in https://github.com/SeldonIO/alibi-detect/pull/359. 

### Summary of changes
1. `score` and `predict` are refactored to ensure state isn't being updated in multiple places. The state is now updated in `_update_state`, which is called by `score`. This means `score` can now be called without `predict` being called.
2. Preprocessing of `x_t` is moved to `_update_state`, meaning preprocessing is now done for `score` and `predict`. 
3. `score` is exposed as a public method of `MMDDriftOnline` and `LSDDDriftOnline`.
4. `x_t` is converted to `np.ndarray` if given as a `list`.
5. Minor rewording of some docstrings.

### CODE REVIEW

File: alibi_detect/cd/base_online.py [hunk 1]
```diff
@@ -99,7 +125,7 @@ def reset(self) -> None:
         "Resets the detector but does not reconfigure thresholds."
         self._initialise()
 
-    def predict(self, x_t: Union[np.ndarray, Any],  return_test_stat: bool = True,
+    def predict(self, x_t: Union[np.ndarray, list],  return_test_stat: bool = True,
```
Comment (arnaudvl) [inline alibi_detect/cd/base_online.py:128]: I don't think that changing form `Any` to `list` as type for `x_t` is correct? We're passing single instances to `predict`, which can be e.g. arrays, string values or custom objects. They are turned into batches (array or list) of size 1 by e.g. `x_t[None, :]` or `[x_t]`.

Comment (ascillitoe) [reply alibi_detect/cd/base_online.py:128]: Thanks, good point. I shall change this back now!

File: alibi_detect/cd/base_online.py [hunk 2]
```diff
@@ -86,6 +89,29 @@ def _configure_thresholds(self):
     def _configure_ref_subset(self):
         pass
 
+    @abstractmethod
+    def _update_state(self, x_t: Union[np.ndarray, list]):
+        pass
+
+    def _preprocess_xt(self, x_t: Union[np.ndarray, list]) -> np.ndarray:
```
Comment (arnaudvl) [inline alibi_detect/cd/base_online.py:96]: @ascillitoe I believe the `list` instead of `Any` type also manifests in other places such as here.

Comment (ascillitoe) [reply alibi_detect/cd/base_online.py:96]: Sorry yes you're right. The previous behavior was for `predict` to take `Any`, and then predict applied the preprocessing such that `score` was safe to accept `np.ndarray` only. With the new `_preprocess_xt` and `_update_state` all this needs a rethink. I shall look again at this now.

Comment (ascillitoe) [reply alibi_detect/cd/base_online.py]: Ok I think I've updated all the relevant `list`'s to `Any`. This includes `predict` in `cd/mmd_online.py` and `cd/lsdd_online.py`, which were currently `list` in the existing code, but I think should also be `Any`?

File: alibi_detect/cd/pytorch/mmd_online.py [hunk 3]
```diff
@@ -167,30 +167,32 @@ def _configure_thresholds(self):
 
         self.thresholds = thresholds
 
-    def score(self, x_t: np.ndarray) -> Union[float, None]:
+    def _update_state(self, x_t: Union[np.ndarray, Any]):
+        self.t += 1
+        x_t = super()._preprocess_xt(x_t)
```
Comment (arnaudvl) [inline alibi_detect/cd/pytorch/mmd_online.py:172]: While I agree that this looks pretty clean, just wanted to make sure we are happy that `_update_state` also contains steps which are not related to state, such as preprocessing of `x_t`?

File: alibi_detect/cd/pytorch/lsdd_online.py [hunk 4]
```diff
@@ -178,26 +178,29 @@ def _configure_ref_subset(self):
             h_init = self.c2s - self.k_xtc.mean(0)  # (Eqn 21)
             lsdd_init = h_init[None, :] @ self.H_lam_inv @ h_init[:, None]  # (Eqn 11)
 
-    def score(self, x_t: np.ndarray) -> Union[float, None]:
+    def _update_state(self, x_t: Union[np.ndarray, Any]):
+        self.t += 1
+        x_t = super()._preprocess_xt(x_t)
```
Comment (arnaudvl) [inline alibi_detect/cd/pytorch/lsdd_online.py:183]: Same comment as in the MMD equivalent: would it make sense to do preprocessing of `x_t` separately from `_update_state` and feed the preprocessed `x_t` to `_update_state`?

File: alibi_detect/cd/tensorflow/mmd_online.py [hunk 5]
```diff
@@ -157,25 +157,27 @@ def _configure_thresholds(self):
 
         self.thresholds = thresholds
 
-    def score(self, x_t: np.ndarray) -> Union[float, None]:
+    def _update_state(self, x_t: Union[np.ndarray, Any]):
+        self.t += 1
+        x_t = super()._preprocess_xt(x_t)
```
Comment (arnaudvl) [inline alibi_detect/cd/tensorflow/mmd_online.py:162]: See comment PyTorch version.

File: alibi_detect/cd/tensorflow/lsdd_online.py [hunk 6]
```diff
@@ -164,27 +164,29 @@ def _configure_ref_subset(self):
             h_init = self.c2s - tf.reduce_mean(self.k_xtc, axis=0)  # (Eqn 21)
             lsdd_init = h_init[None, :] @ self.H_lam_inv @ h_init[:, None]  # (Eqn 11)
 
-    def score(self, x_t: np.ndarray) -> Union[float, None]:
+    def _update_state(self, x_t: Union[np.ndarray, Any]):
+        self.t += 1
+        x_t = super()._preprocess_xt(x_t)
```
Comment (arnaudvl) [inline alibi_detect/cd/tensorflow/lsdd_online.py:169]: See comment PyTorch version.

File: alibi_detect/cd/pytorch/lsdd_online.py (hunk 4 above)
Comment (ascillitoe) [reply alibi_detect/cd/pytorch/lsdd_online.py:183]: Sure, we can certainly separate these. My thinking behind calling `_preprocess_xt` in `_update_state` was that essentially we are updating state with `x_t`, so it does make some sense to have it inside `_update_state` (as we're only preprocessing `x_t` in order to update the state). But I suppose thinking about more complex handling of state in the future (loading state etc), perhaps it is clearer for us to separate out instead.

**Review by ojcobb — APPROVED** [review]
LGTM, nice one! Thanks for catching some of my outdated type hints also.

**Review by arnaudvl — APPROVED** [review]

**Pull request merged**